summingbird-storm/src/main/scala/com/twitter/summingbird/storm/collector/AggregatorOutputCollector.scala

```diff
@@ -0,0 +1,122 @@
+package com.twitter.summingbird.storm.collector
+
+import com.twitter.algebird.Semigroup
+import com.twitter.summingbird.online.executor.KeyValueShards
+import com.twitter.summingbird.online.option.SummerBuilder
+import backtype.storm.spout.SpoutOutputCollector
+import com.twitter.algebird.util.summer.AsyncSummer
+import com.twitter.util.{ Await, Future, Time }
+import scala.collection.mutable.{ Map => MMap }
+import scala.collection.mutable.{ MutableList => MList }
+import scala.collection.{ Map => CMap }
+import scala.collection.JavaConverters._
+import java.util.{ List => JList }
+
+/**
+ *
+ * AggregatorOutputCollector is a wrapper around the SpoutOutputCollector.
+ * AsyncSummer is used to aggregate the tuples.
+ * Different streams have seperated aggregators and caches.
+ *
+ */
+class AggregatorOutputCollector[K, V: Semigroup](
+    in: SpoutOutputCollector,
+    summerBuilder: SummerBuilder,
+    summerShards: KeyValueShards) extends SpoutOutputCollector(in) {
+
+  // Map keeps track of summers corresponding to streams.
+  private val spoutCaches = MMap[String, AsyncSummer[(K, V), Map[K, V]]]()
+
+  // The Map keeps track of batch of aggregated tuples' messageIds. It also has a stream level tracking.
+  private val streamMessageIdTracker = MMap[String, MMap[Int, MList[Object]]]()
+
+  def timerFlush() = {
+    /*
+    This is a flush called from the nextTuple() of the spout.
+    The timerFlush is triggered with tick frequency from the spout.
+     */
+    spoutCaches.foreach {
+      case (stream, cache) =>
+        val tupsOut = cache.tick.map { convertToSummerInputFormat(_) }
+        emitData(tupsOut, stream)
+    }
+  }
+
+  private def convertToSummerInputFormat(flushedCache: CMap[K, V]): CMap[Int, CMap[K, V]] =
+    flushedCache.groupBy { case (k, _) => summerShards.summerIdFor(k) }
+
+  /*
+    The method is invoked to handle the flushed cache caused by
+    exceeding the memoryLimit, which is called within add method.
+   */
+  private def emitData[K, V](cache: Future[TraversableOnce[(Int, CMap[K, V])]], streamId: String): JList[Integer] = {
+    val flushedTups = Await.result(cache)
```

Comment:
The way AsyncSummer behaves is that it will return empty result most of the time and non-empty results only when it flushes. We should short circuit for empty result to avoid processing below in those cases, e.g. creation of ArrayList. Also blocking on future for every single event seems like a performance problem, although I am not sure how expensive await is.